lib/Basics/FutureSharedLock.h

```diff
@@ -0,0 +1,246 @@
+    });
+  }
+
+  void unlock() { _sharedState->unlock(); }
+
+ private:
+  struct Node {
+    explicit Node(bool exclusive) : exclusive(exclusive) {}
+
+    Promise<LockGuard> promise;
+    bool exclusive;
+    bool scheduled{false};
+  };
+
+  struct SharedState : std::enable_shared_from_this<SharedState> {
+    explicit SharedState(Scheduler& scheduler) : _scheduler(scheduler) {}
+
+    template<class Func>
+    FutureType asyncLockExclusive(Func blockedFunc) {
+      std::lock_guard lock(_mutex);
+      if (_lockCount == 0) {
+        TRI_ASSERT(_queue.empty());
+        ++_lockCount;
+        _exclusive = true;
+        return {LockGuard(this)};
+      }
+
+      auto it = insertNode(true);
+      blockedFunc(it);
+      return (*it)->promise.getFuture();
+    }
+
+    template<class Func>
+    FutureType asyncLockShared(Func blockedFunc) {
+      std::lock_guard lock(_mutex);
+      if (_lockCount == 0 || (!_exclusive && _queue.empty())) {
+        ++_lockCount;
+        _exclusive = false;
+        return {LockGuard(this)};
+      }
+
+      auto it = insertNode(false);
+      blockedFunc(it);
+      return (*it)->promise.getFuture();
+    }
+
+    void unlock() {
+      std::lock_guard lock(_mutex);
+      TRI_ASSERT(_lockCount > 0);
+      if (--_lockCount == 0 && !_queue.empty()) {
+        // we were the last lock holder -> schedule the next node
+        auto& node = _queue.back();
+        _lockCount = 1;
+        _exclusive = node->exclusive;
+        scheduleNode(node);
+        _queue.pop_back();
+
+        // if we are in shared mode, we can schedule all following shared
+        // nodes
+        if (!_exclusive) {
+          while (!_queue.empty() && !_queue.back()->exclusive) {
+            ++_lockCount;
+            scheduleNode(_queue.back());
+            _queue.pop_back();
+          }
+        }
+      }
+    }
+
+    auto insertNode(bool exclusive) {
+      _queue.emplace_front(std::make_shared<Node>(exclusive));
+      return _queue.begin();
+    }
+
+    void removeNode(
+        typename std::list<std::shared_ptr<Node>>::iterator queueIterator) {
+      TRI_ASSERT(!_queue.empty());
+      TRI_ASSERT(_lockCount > 0);
+      if (std::addressof(_queue.back()) == std::addressof(*queueIterator)) {
+        // we are the first entry in the list -> remove ourselves and check if
+        // we can schedule the next node(s)
+        _queue.pop_back();
+        if (!_exclusive) {
+          while (!_queue.empty() && !_queue.back()->exclusive) {
+            ++_lockCount;
+            scheduleNode(_queue.back());
+            _queue.pop_back();
+          }
+        }
+      } else {
+        // we are somewhere in the middle -> just remove ourselves
+        _queue.erase(queueIterator);
+      }
+    }
+
+    void scheduleExclusive(Node* node) { scheduleNode(node); }
+
+    void scheduleShared(Node* leader) { scheduleNode(leader); }
+
+    void scheduleNode(std::shared_ptr<Node> node) {
```

Comment:
We don't need to copy shared_ptr here
```suggestion
    void scheduleNode(Node& node) {
```